### PR TITLE
flyctl machine status

### PR DESCRIPTION
**Pull request by rugwirobaker**
`flyctl machine status` so far:
Screen Shot 2022-02-07 at 16 09 27

### CODE REVIEW

**Review by shortdiv — APPROVED** [review]
Looks good to me!

**Comment by rugwirobaker** [comment]
Update: @shortdiv what do you think of a diff field here?
Screen Shot 2022-02-10 at 18 13 51

**Comment by shortdiv** [comment]
I don't think the diff object really makes sense, and it's apparently difficult to retrieve from the database into graphql. Instead of showing diff, we should show machine event types, from start/exit/destroy i.e.`exit_code`, `oom_killed` and `requested_stop`, you can see available fields here: https://github.com/superfly/web/blob/master/app/graphql/types/machine_event_exit.rb

File: cmd/presenters/machine_events.go [hunk 1]
```diff
@@ -18,18 +19,26 @@ func (p *MachineEvents) APIStruct() interface{} {
...
+		if event.ExitCode != nil {
+			exitCode = fmt.Sprintf("%d", *event.ExitCode)
+		}
+
 		out = append(out, map[string]string{
+			"Id":        event.ID,
 			"Kind":      event.Kind,
 			"Timestamp": event.Timestamp.Format(time.RFC3339),
-			// "Message":   event.Message,
+			`Exit Code`: exitCode,
```
Comment (shortdiv) [inline cmd/presenters/machine_events.go:41]: [nit] Same as above, consistency on using "" vs ``for strings

File: api/types.go [hunk 2]
```diff
@@ -1231,7 +1232,43 @@ type DeleteOrganizationMembershipPayload struct {
...
+	Timestamp time.Time
+	ExitCode  *int
+	OOMKilled bool
+	// Body      map[string]interface{}
+}
+
+func (e *MachineEvent) UnmarshalJSON(data []byte) error {
+	raw := map[string]interface{}{}
+
+	if err := json.Unmarshal(data, &raw); err != nil {
```
Comment (shortdiv) [inline api/types.go:1246]: I think unmarshalling the data directly into the struct would save a lot of trouble when it comes to type assertions, something like:

```
var raw MachineEvent
if err := json.Unmarshal(data, &raw); err != nil {
  return err
}
```

Which will then allow you to just directly check for data like raw.ID etc

File: api/types.go [hunk 3]
```diff
@@ -1231,7 +1232,43 @@ type DeleteOrganizationMembershipPayload struct {
...
+		return err
+	}
+
+	e.ID = raw["id"].(string)
+	delete(raw, "id")
```
Comment (shortdiv) [inline api/types.go:1251]: [nit] What's the reason for deleting data from the map?

File: api/types.go (hunk 2 above)
Comment (rugwirobaker) [reply api/types.go:1246]: MachineEvent takes more than one form such as `MachineEventStart`, `MachineEventExit`... and they don't share all the fields and so I must be more generic and manually check if the fields exists before trying to set them.

File: api/types.go (hunk 3 above)
Comment (rugwirobaker) [reply api/types.go:1251]: Nothing special, just a way of keeping only the data that has not yet been set unto a field.

Comment (shortdiv) [reply api/types.go:1251]: Ah ok, I think this is probably stylistic, but I don't see the point of it if the object gets garbage collected anyway once the function returns

File: api/types.go (hunk 2 above)
Comment (shortdiv) [reply api/types.go:1246]: On second thought I think a case switch could work, something like this: https://play.golang.com/p/49xJe_HysXb

File: api/types.go [hunk 4]
```diff
@@ -1235,40 +1235,35 @@ type MachineEvent struct {
...
 func (e *MachineEvent) UnmarshalJSON(data []byte) error {
-	raw := map[string]interface{}{}
+	var genericMachineEvent MachineEvent
+	var machineStop MachineEventStop
 
-	if err := json.Unmarshal(data, &raw); err != nil {
+	if err := json.Unmarshal(data, &genericMachineEvent); err != nil {
 		return err
 	}
+	machineBodyinBytes, _ := json.Marshal(genericMachineEvent.Metadata)
```
Comment (rugwirobaker) [inline api/types.go:1253]: This right here causes a stack overflow:
```bash
 go run . machine status 71d5c016 --app <app-name>
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0x14020700550 stack=[0x14020700000, 0x14040700000]
fatal error: stack overflow

```

**Pull request merged**